src/server/frontend_wayland/xdg_activation_v1.cpp

```diff
@@ -0,0 +1,312 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "xdg_activation_v1.h"
+#include "wl_surface.h"
+#include "wl_seat.h"
+#include "mir/main_loop.h"
+#include "mir/scene/surface.h"
+#include "mir/shell/shell.h"
+#include "mir/shell/focus_controller.h"
+#include "mir/log.h"
+#include <random>
+#include <chrono>
+
+namespace mf = mir::frontend;
+namespace mw = mir::wayland;
+namespace msh = mir::shell;
+
+namespace mir::frontend
+{
+/// Time in milliseconds that the compositor will wait before invalidating a token
+auto constexpr TIMEOUT_MS = std::chrono::milliseconds(3000);
+
+std::string generate_token()
+{
+    std::string str("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz");
+    std::random_device rd;
+    std::mt19937 generator(rd());
+    std::shuffle(str.begin(), str.end(), generator);
+    return str.substr(0, 32);
+}
```

Comment:
This looks somewhat suspect (limited range of characters, each character can only occur once).

Reply:
Yeah this is a good candidate to use `uuid`